Fix max/min.reduction_with_dim opinfo test for bool tensors

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109359
* __->__ #109264
* #109165

The tests were failing because the input tensors were getting
incorrectly promoted to int64, due to `transform_args` incorrectly
considering the `dim` argument when determining the type to promote to.

It doesn't seem like type promotion is necessary in general for max and
min, so I've switched them to `type_promotion_kind=None`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

int3 committed:
Fix max/min.reduction_with_dim opinfo test for bool tensors

The tests were failing because the input tensors were getting
incorrectly promoted to int64, due to `transform_args` incorrectly
considering the `dim` argument when determining the type to promote to.

I've fixed this by adding an option to explicitly specify which args
should be considered when doing type promotion.

[ghstack-poisoned]

int3 committed:
Update on "Fix max/min.reduction_with_dim opinfo test for bool tensors"


The tests were failing because the input tensors were getting
incorrectly promoted to int64, due to `transform_args` incorrectly
considering the `dim` argument when determining the type to promote to.

It doesn't seem like type promotion is necessary in general for max and
min, so I've switched them to `type_promotion_kind=None`.

[ghstack-poisoned]

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -4545,7 +4545,7 @@ def reduce_any(x, dim=None, keepdim=False):
 
 
-@register_lowering(aten.max)
+@register_lowering(aten.max, type_promotion_kind=None)
```
What was the type promotion that was happening previously ? It looks like only the tensor inputs should be used in type promotion: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L181. So i'm curious exactly what was happening here

int3 replied:
Yes, only tensors are candidates for promotion, but when considering the dtype to promote to, `dim` gets considered as well since it is a Number: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L187-L190, and the resulting dtype is an int64.

Presumably this is done because some number parameters are meant to be added or multiplied with the tensor

eellison reviewed (commented):
looks good - want to understand what was happening

mlazos commented on torch/_inductor/lowering.py:
```diff
@@ -4556,7 +4556,7 @@ def reduce_max(x, dim=None, keepdim=False):
 
 
-@register_lowering(aten.min)
+@register_lowering(aten.min, type_promotion_kind=None)
```
Will this match the type promotion behavior in eager? I feel like eager might promote

int3 replied:
Do you have a particular case in mind where promotion would happen? Promotion doesn't happen for the following cases, at least:

```
In [13]: torch.max(torch.zeros((2,2), dtype=torch.bool), dim=1)[0].dtype
Out[13]: torch.bool

In [14]: torch.max(torch.zeros((2,2), dtype=torch.float16), dim=1)[0].dtype
Out[14]: torch.float16

In [15]: torch.max(torch.zeros((2,2), dtype=torch.int32), dim=1)[0].dtype
Out[15]: torch.int32
```

eellison reviewed (approved):
looks good - maybe in a follow up it's worth auditing our other registrations, or adding logic so that we don't include `dim` as a type-promoting arg

int3 commented:
@pytorchbot merge

int3 committed:
Update on "Fix max/min.reduction_with_dim opinfo test for bool tensors"


The tests were failing because the input tensors were getting
incorrectly promoted to int64, due to `transform_args` incorrectly
considering the `dim` argument when determining the type to promote to.

It doesn't seem like type promotion is necessary in general for max and
min, so I've switched them to `type_promotion_kind=None`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 committed:
Update on "Fix max/min.reduction_with_dim opinfo test for bool tensors"


The tests were failing because the input tensors were getting
incorrectly promoted to int64, due to `transform_args` incorrectly
considering the `dim` argument when determining the type to promote to.

It doesn't seem like type promotion is necessary in general for max and
min, so I've switched them to `type_promotion_kind=None`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented:
@pytorchbot merge
